Core/src/Geometry/CuboidVolumeStack.cpp

```diff
@@ -0,0 +1,893 @@
+#include "Acts/Utilities/AxisDefinitions.hpp"
+#include "Acts/Utilities/Logger.hpp"
+
+#include <algorithm>
+#include <array>
+#include <cstddef>
+#include <iomanip>
+#include <memory>
+#include <sstream>
+#include <utility>
+
+namespace Acts {
+
+struct CuboidVolumeStack::VolumeTuple {
+  Volume* volume{};
+  const CuboidVolumeBounds* bounds{};
+  std::shared_ptr<CuboidVolumeBounds> updatedBounds{};
+  Transform3 localTransform = Transform3::Identity();
+  Transform3 globalTransform = Transform3::Identity();
+
+  bool transformDirty = false;
+
+  VolumeTuple(Volume& volume_, const Transform3& groupTransform)
+      : volume{&volume_},
+        localTransform{groupTransform.inverse() * volume_.transform()},
+        globalTransform{volume_.transform()} {
+    bounds = dynamic_cast<const CuboidVolumeBounds*>(&volume_.volumeBounds());
+    assert(bounds != nullptr);
+    updatedBounds = std::make_shared<CuboidVolumeBounds>(*bounds);
+  }
+
+  double mid(AxisDirection direction) const {
+    return localTransform.translation()[static_cast<std::size_t>(direction)];
+  }
+  double halfLength(AxisDirection direction) const {
+    return updatedBounds->get(
+        static_cast<CuboidVolumeBounds::BoundValues>(direction));
+  }
+  double min(AxisDirection direction) const {
+    return mid(direction) - halfLength(direction);
+  }
+  double max(AxisDirection direction) const {
+    return mid(direction) + halfLength(direction);
+  }
+
+  void set(
+      std::initializer_list<std::pair<CuboidVolumeBounds::BoundValues, double>>
+          keyValues) {
+    updatedBounds->set(keyValues);
+  }
+
+  void setLocalTransform(const Transform3& transform,
+                         const Transform3& groupTransform) {
+    localTransform = transform;
+    globalTransform = groupTransform * localTransform;
+    transformDirty = true;
+  }
+
+  void commit(const Logger& logger) {
+    // make a copy so we can't accidentally modify in-place
+    auto copy = std::make_shared<CuboidVolumeBounds>(*updatedBounds);
+
+    std::optional<Transform3> transform = std::nullopt;
+    if (transformDirty) {
+      transform = globalTransform;
+    }
+
+    volume->update(std::move(updatedBounds), transform, logger);
+    bounds = copy.get();
+    updatedBounds = std::move(copy);
+    transformDirty = false;
+  }
+};
+
+CuboidVolumeStack::CuboidVolumeStack(std::vector<Volume*>& volumes,
+                                     AxisDirection direction,
+                                     AttachmentStrategy strategy,
+                                     ResizeStrategy resizeStrategy,
+                                     const Logger& logger)
+    : Volume(initialVolume(volumes)),
+      m_dir(direction),
+      m_resizeStrategy(resizeStrategy),
+      m_volumes(volumes) {
+  auto dirIdx = static_cast<std::size_t>(m_dir);
+  m_dirOrth1 = static_cast<AxisDirection>((dirIdx + 1) % 3);
+  m_dirOrth2 = static_cast<AxisDirection>((dirIdx + 2) % 3);
```

Comment:
I see what this is trying to accomplish. I'm not entirely sure, if we should rely on the internal numbering inside `AxisDirection` to perform this rotation. Could this be spelled out with regular control flow like a `switch`? It shouldn't be a lot of cases, should it?